### ISSUE TITLE
`torch.where` produces nan in backward pass for differentiable forward pass

### ISSUE THREAD

**Issue by bauermeisterh** [pytorch/pytorch#68425]
## To Reproduce

In the example below, `torch.where` properly executes a differentiable forward pass but fails for calculating the correct gradient value:
```
a = torch.tensor(100., requires_grad=True)
b = torch.where(a <= 0, torch.exp(a), 1 + a)
b.backward()

print(b)
print(a.grad)
```
outputs
```
tensor(101., grad_fn=<SWhereBackward>)
tensor(nan)
```

## Expected behavior

The expected gradient in the example above is `1.` for `a.grad`.

## Environment

PyTorch version: 1.9.1
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.9.4 (default, Apr  9 2021, 16:34:09)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-5.11.0-40-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce GTX 1050 Ti
GPU 1: TITAN Xp

Nvidia driver version: 460.91.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.1
[pip3] numpy==1.20.1
[pip3] numpydoc==1.1.0
[pip3] torch==1.9.1
[pip3] torchaudio==0.9.0a0+a85b239
[pip3] torchvision==0.10.1
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.1.74              h6bb024c_0    nvidia
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.2.0           h06a4308_296  
[conda] mkl-service               2.3.0            py39h27cfd23_1  
[conda] mkl_fft                   1.3.0            py39h42c9631_2  
[conda] mkl_random                1.2.1            py39ha9443f7_2  
[conda] mypy_extensions           0.4.1            py39h06a4308_0  
[conda] numpy                     1.20.1           py39h93e21f0_0  
[conda] numpy-base                1.20.1           py39h7d8b39e_0  
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1  
[conda] pytorch                   1.9.1           py3.9_cuda11.1_cudnn8.0.5_0    pytorch
[conda] torchaudio                0.9.1                      py39    pytorch
[conda] torchvision               0.10.1               py39_cu111    pytorch

cc @brianjo @mruberry @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by gchanan** [comment]
@albanD this comes up quite a bit -- do we have documentation / a note we can point to for when people file these issues?

**Comment by albanD** [comment]
We do have other issues where this is discussed: https://github.com/pytorch/pytorch/issues/23156
Otherwise I think the presentation on masked Tensor is a good explanation, cc @cpuhrsch is that ready to be shared more broadly?

**Comment by cpuhrsch** [comment]
It'd be useful to get prototype feedback on [maskedtensor](https://github.com/cpuhrsch/maskedtensor), and of course anyone is welcome to use it, but I wouldn't want to put someone through the potential significant pain of building something stable on top of it. It'll likely address this issue, but it'll also still change a bit until a stable release. I don't want to commit to a timeline here on the spot, because it's not clear.

**Comment by gchanan** [comment]
My point is more -- having a well documented place to reference is helpful here because it's more clear and discoverable than ongoing issue discussions or presentations.  It's the same situation as https://pytorch.org/docs/master/notes/numerical_accuracy.html#batched-computations-or-slice-computations, which have been helpful to explain what's going on.

**Comment by UlrichFreitag** [comment]
Is there any news on this issue? :smiley:

**Comment by nikitaved** [comment]
Can we just modify the forward to mask out relevant parts?

**Comment by nikitaved** [comment]
@cpuhrsch , yes! But, apparently, this does not work either...

**Comment by cpuhrsch** [comment]
@nikitaved The reason this doesn't work is because nan's [propagate through multiplication](https://en.wikipedia.org/wiki/NaN#Operations_generating_NaN).

**Comment by nikitaved** [comment]
@cpuhrsch , I still do not understand where nans creep from in this particular case unless autograd is generating empty/nan parts.
```python
In [1]: import torch

In [2]: a = torch.tensor([100., -10], requires_grad=True)
   ...: mask = (a.detach() <= 0)
   ...: b = mask * torch.exp(a) + (1 + a) * (~mask)

In [3]: b.sum().backward()

In [4]: a.grad
Out[4]: tensor([       nan, 4.5400e-05])

```

**Comment by cpuhrsch** [comment]
In this case ```mask * torch.exp(a)``` will have a NaN. And yes, look at the derivatives for each operation as well to gain a sense for what's going on here.

**Comment by Lezcano** [comment]
I always thought that the backward for `torch.where(mask, x, y)` could be implemented by doing:
`torch.masked_scatter(grad_y, torch.masked_select(grad_x, mask))`
And it may be even faster?

**Comment by nikitaved** [comment]
Or that. I still believe we can solve this issues with the tools we have at our hands.

**Comment by cpuhrsch** [comment]
@Lezcano @nikitaved - but maintaining an additional mask, until it's not required anymore, seems more efficient than these scatter and select operations. The gradient that's coming in torch.where also becomes a lot simpler. We don't even need to call where anymore and fill the places with zeros that weren't used in the forward. We just keep the mask.

Sure we need additional memory (one bit per element at best), but if this is a major concern the user can revert back to our usual operations and find a different workaround. For example, why apply torch.exp to a value that high to begin with.

```
>>> import torch
>>> a = torch.tensor(100., requires_grad=True)
>>> mask = a <= 0
>>> b = torch.exp(mask * a) * mask + (~mask) * (1 + a)
>>>
>>> b.backward()
>>>
>>> print(b)
tensor(101., grad_fn=<AddBackward0>)
>>> print(a.grad)
tensor(1.)
```

**Comment by ngimel** [comment]
@cpuhrsch why are you doing `exp(mask*a)`? shouldn't it be `exp(a) * mask`? If I'm doing `exp(a)* mask` I'm getting a nan gradient, as expected.

**Comment by Lezcano** [comment]
My comment didn't make much sense, sorry. And yes, I don't see how this could be mitigated with what we currently have.

**Comment by nikitaved** [comment]
@cpuhrsch , unfortunately, this approach will not work for non-linear functions of which exp is the example. But even with masking there is still a question of propagating through `inf * 0`.

**Comment by cpuhrsch** [comment]
```exp(mask * a)``` will map masked-out elements to 1 and apply exp to those less than or equal to 0. 
```exp(mask * a) * mask``` further replaces those ones (masked-out elements) with zero.
Then you add ```~mask * (1 + a)``` which will replace those zeros (or masked out values) with ```1 + a```.

So in the end you apply exp to those values of ```a``` less than or equal to 0 and + 1 to those greater than 0.

The reason I'm not getting a nan gradient here, is because I never get an inf, since ```exp``` is applied to values less than or equal to zero via ```mask * a```.

Doesn't ```torch.where(a <= 0, torch.exp(a), 1 + a)``` do the same thing?

**Comment by nikitaved** [comment]
Ah, good! Yes, it will work with exp. But if we replace it with log, then this solution will not work as well.

**Comment by cpuhrsch** [comment]
@nikitaved - yes, but the point is that the user can get it done, if they really want to. Otherwise we should use MaskedTensor.

**Comment by nikitaved** [comment]
I see, then what shall we do with `where`? I think we can fix it, given how popular it is, albeit the solution could get more involved with, say, a custom kernel that is doing the mul(tensor, bool) without any explicit multiplications.

**Comment by cpuhrsch** [comment]
@nikitaved - it'll stay like it is and nans propagate. That's it. We shouldn't make things like ```inf * 0 == 0```. It breaks IEEE754 simply because we don't have the metadata available to track invalid elements. If a user really needs to filter the creation of nans they'll have to do so earlier on. For your log example consider

```torch.log(a.masked_fill(~mask, 1)) * mask + (~mask) * (1 + a)```

**Comment by nikitaved** [comment]
I do not think we break the standard here, as `where` is not meant to do any explicit multiplications, it just chooses an element based on a condition,

**Comment by ngimel** [comment]
@nikitaved backward for `where` is not a problem here, it sends `0` gradient to `torch.exp(a)` (self) and `1` gradient to 1+a (other) (or, in your example with inverted mask, `1` gradient to `torch.exp(a)` and `0` gradient to `1+a`. When those gradients are further backpropagated to `a`, 1*inf multiplication happens that results in `nan`, but `where` backward is not to blame.

**Comment by nikitaved** [comment]
But `1 * inf` should not be nan, it is `inf`, and the first example in my previous post is doing it right.

**Comment by ngimel** [comment]
Sorry, `0*inf` leads to `nan`, when `exp(a)` element is selected, then it's `1*inf` and it's `inf`

**Comment by nikitaved** [comment]
@ngimel , do we agree that we can eliminate any explicit multiplications `0 * inf` and instead do a "logical" multiplication in this case? Or rather, that it is a problem and we could fix that?

**Comment by ngimel** [comment]
How can we eliminate `0*inf` multiplication in the backward of `exp(a)`? It requires an actual (and not logical) multiplication

**Comment by nikitaved** [comment]
In `where(exp, exp)` only one branch gets the gradient, the other one gets a zero tensor with entries selected from the inverted mask, which is all zeros, so no elements are going to get selected and the grad propagated to exp is will be zero.

**Comment by ngimel** [comment]
Exactly, the grad propagated to `exp` is 0 (and `where` does that), but then when `grad` is propagated from `exp(a)` back to `a) (because we cannot reason that the backpropagation should stop at that point) `0 * inf` multiplication happens. If there's an `inf` input to `where`, everything is fine (it gets 0 grad):
```
In [21]: b=torch.tensor(math.inf, requires_grad=True)

In [22]: a=torch.tensor(100., requires_grad=True)

In [23]: c=torch.where(a<0, b, a)

In [24]: c.backward()

In [25]: b.grad
Out[25]: tensor(0.)

In [26]: a.grad
Out[26]: tensor(1.)

```

**Comment by nikitaved** [comment]
I see your point, but what if an un-defined gradients gets propagated to `exp(a)`? In such a case `zeros_like(a)` get propagated back to `a`.

**Comment by ngimel** [comment]
But we don't have a concept of "undefined grad"

**Comment by ngimel** [comment]
That's for multi-output functions that sometimes get gradients only for one of the outputs. In this simple case it may seem that if only we said that `exp(a)` received undefined gradient, everything would be fine, but if `a` was a multi-element tensor, it should receive a gradient where some elements are non-zero (they were selected) and some elements are 0 because they are masked out, and there's no way to represent it with just a single undefined gradient.

**Comment by nikitaved** [comment]
I see your point, @ngimel . Thanks for this great discussion, @cpuhrsch and @ngimel , that was quite illuminating!

**Comment by cpuhrsch** [comment]
@nikitaved - To conclude the conversation, do you want to give [MaskedTensor](https://github.com/cpuhrsch/maskedtensor) a shot?

**Comment by nikitaved** [comment]
@cpuhrsch , most certainly!

**Comment by soulitzer** [comment]
Duplicate of https://github.com/pytorch/pytorch/issues/4132

**Comment by Lucas-vdr-Horst** [comment]
I just encountered this issue as well. Would it be possible to add a note to the `torch.where` documentation?
